### PR TITLE
feat(finance): Black-Scholes pricer+Greeks+IV, Kelly sizing, StandardScaler NaN fix

### PR DESCRIPTION
**Pull request by ooples**
BlackScholes<T> (price+delta/gamma/vega/theta/rho+IV), KellyCriterion<T> (discrete/continuous/fractional/from-returns), and a StandardScaler NaN-on-constant-column fix. 23 unit tests pass (net10+net471) vs textbook values.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Black‑Scholes pricer with Greeks and implied volatility solver
  * Kelly‑criterion sizing (discrete/continuous/ fractional)
  * Markowitz mean‑variance optimizer
  * Risk metrics: Sharpe, Sortino, Calmar
  * Pairwise ranking loss with optional tail‑weighting
  * NDCG ranking metric

* **Bug Fixes**
  * Guarded against tiny negative variance in standard scaler

* **Improvements**
  * Reduced allocations in matrix decomposition for faster decompositions
  * Improved ONNX input wiring for broader model compatibility

* **Tests**
  * Added unit tests covering new finance, ranking, and ONNX input behavior

### CODE REVIEW

File: src/Finance/Options/BlackScholes.cs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+/// actually needs. Feed a GARCH / forecast volatility in as <c>volatility</c>.
+/// </para>
+/// <para><b>For Beginners:</b> The Black-Scholes formula gives the fair price of a European call or put
+/// option from five inputs: the current price (spot), the strike, time to expiry (in years), the
+/// risk-free interest rate, and the volatility. The "Greeks" are its sensitivities — delta (price vs.
+/// spot), gamma (delta vs. spot), vega (price vs. volatility), theta (price vs. time), rho (price vs.
+/// rate) — which traders use to size positions and hedge risk.</para>
+/// </remarks>
+/// <typeparam name="T">Numeric type (float/double).</typeparam>
+public static class BlackScholes<T>
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/Finance/Options/BlackScholes.cs]: The facade rule applies to ML models built via the AiModelBuilder/AiModelResult pipeline. These are standalone, stateless quantitative-finance utilities (closed-form pricing, position sizing, risk ratios) — not trainable models — so routing them through the model-builder facade does not fit. Keeping them public as a deliberate domain API, consistent with the rest of src. The concrete validation/correctness issues raised in the sibling threads are fixed in this push.

File: src/Finance/Portfolio/KellyCriterion.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+/// payoff odds) and continuous (mean / variance of returns) — plus fractional Kelly, which is what
+/// practitioners actually use because full Kelly is famously over-aggressive.
+/// </para>
+/// <para><b>For Beginners:</b> Kelly tells you what fraction of your capital to put on a trade to grow
+/// wealth fastest over many trades. Bet more than Kelly and you risk ruin; bet a fraction of it
+/// ("half-Kelly") for much lower volatility at a small growth cost. A negative Kelly means "no edge —
+/// don't take the trade."</para>
+/// </remarks>
+/// <typeparam name="T">Numeric type (float/double).</typeparam>
+public static class KellyCriterion<T>
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/Finance/Portfolio/KellyCriterion.cs]: The facade rule applies to ML models built via the AiModelBuilder/AiModelResult pipeline. These are standalone, stateless quantitative-finance utilities (closed-form pricing, position sizing, risk ratios) — not trainable models — so routing them through the model-builder facade does not fit. Keeping them public as a deliberate domain API, consistent with the rest of src. The concrete validation/correctness issues raised in the sibling threads are fixed in this push.

File: src/Finance/Portfolio/MarkowitzOptimizer.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+/// </para>
+/// <para><b>For Beginners:</b> "Mean-variance" optimization picks how much to put in each asset by
+/// trading off expected return against risk (variance), using how the assets move together (the
+/// covariance matrix). The <i>minimum-variance</i> portfolio is the lowest-risk fully-invested mix and
+/// ignores returns entirely. The <i>tangency</i> portfolio is the mix with the best risk-adjusted return
+/// (highest Sharpe ratio) given a risk-free rate. <i>Target-return</i> finds the lowest-risk mix that
+/// hits a return you specify. All three allow negative weights (short selling) and sum to 100%.</para>
+/// </remarks>
+/// <typeparam name="T">Numeric type (float/double).</typeparam>
+public static class MarkowitzOptimizer<T>
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/Finance/Portfolio/MarkowitzOptimizer.cs]: The facade rule applies to ML models built via the AiModelBuilder/AiModelResult pipeline. These are standalone, stateless quantitative-finance utilities (closed-form pricing, position sizing, risk ratios) — not trainable models — so routing them through the model-builder facade does not fit. Keeping them public as a deliberate domain API, consistent with the rest of src. The concrete validation/correctness issues raised in the sibling threads are fixed in this push.

File: src/Finance/Risk/RiskRatios.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+/// AiDotNet exposes Sharpe on portfolio optimizers and trading agents, but Sortino (downside-only) and
+/// Calmar (return-over-max-drawdown) were missing as standalone, series-based utilities. These let any
+/// forecaster/strategy backtest be scored on risk-adjusted return without a portfolio or agent object.
+/// </para>
+/// <para><b>For Beginners:</b> Sharpe divides average return by total volatility; Sortino only counts
+/// <i>downside</i> volatility (it doesn't punish upside swings); Calmar divides annualized return by the
+/// worst peak-to-trough drawdown. Higher is better for all three.</para>
+/// </remarks>
+/// <typeparam name="T">Numeric type (float/double).</typeparam>
+public static class RiskRatios<T>
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/Finance/Risk/RiskRatios.cs]: The facade rule applies to ML models built via the AiModelBuilder/AiModelResult pipeline. These are standalone, stateless quantitative-finance utilities (closed-form pricing, position sizing, risk ratios) — not trainable models — so routing them through the model-builder facade does not fit. Keeping them public as a deliberate domain API, consistent with the rest of src. The concrete validation/correctness issues raised in the sibling threads are fixed in this push.

File: tests/AiDotNet.Tests/UnitTests/Finance/KellyCriterionTests.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        Assert.Equal(0.0, KellyCriterion<double>.Discrete(0.6, 0.0), 6); // degenerate odds
+    }
+
+    [Fact]
+    public void Continuous_is_mean_over_variance()
+    {
+        // f* = μ / σ² = 0.1 / 0.04 = 2.5
+        Assert.Equal(2.5, KellyCriterion<double>.Continuous(0.1, 0.04), 6);
+        Assert.Equal(0.0, KellyCriterion<double>.Continuous(0.1, 0.0), 6); // zero variance guard
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tests/UnitTests/Finance/KellyCriterionTests.cs:30]: Optional nitpick — the existing tests already cover the no-edge (returns 0) and edge cases; a negative-expected-value case adds little beyond the winProbability/winLossRatio guards now enforced. Leaving as-is.
